Confidence Scoring

s-jse opened the pull request:
This PR adds support for calibration of neural networks (only `TransformerSeq2Seq` for now). Specifically, it adds the code to:
- Save features that are necessary for calibration. For instance, token-level and sequence level probabilities/logits before and after applying dropout (applying dropout as in https://arxiv.org/pdf/1506.02142.pdf).
- Train a boosted tree classifier using these features to predict whether or not an example is parsed correctly. The classifier's score (a number between 0 and 1) is used as a measure of confidence.
- Optionally, plot precision-recall and other useful curves.

s-jse committed:
Plot precision-recall curve of the calibrator

s-jse committed:
Add ConfidenceOutput with several basic features

s-jse committed:
Save confidence features to a file

s-jse committed:
Add code to train a boosted tree calibrator

s-jse committed:
Add pass rate vs. accuracy plot

s-jse committed:
Add logits without dropout as a confidence feature

s-jse committed:
Add confidence estimation to the main prediction code

s-jse committed:
Calculate AUC for logprob

s-jse committed:
Add threshold plot

s-jse committed:
Add calibration command to genienlp

s-jse committed:
Unify logprob with other calibration features

s-jse committed:
Minor change

s-jse committed:
Refactor code that writes predictions to file

s-jse committed:
Delete LSTM

s-jse committed:
Fix bugs introduced by rebase

s-jse committed:
Add LSTM

s-jse committed:
Add more confidence features

s-jse committed:
Remove unnused command line arguments

s-jse committed:
Refactor code; save normalization function after training

s-jse committed:
Add sequence probabilities as confidence features

s-jse committed:
Change confidence featurizers to lambda functions

s-jse committed:
Move all feature calculations to `calibrate.py`

s-jse committed:
Add input length as a confidence feature

s-jse committed:
Log when loading confidence estimator

s-jse committed:
Fix bug in generation arguments

s-jse committed:
Add CEV as a confidence feature

s-jse committed:
Add variance of logprob as confidence feature

s-jse committed:
Add required packages for calibration

s-jse committed:
Document calibrator related commands

s-jse committed:
Add oracle confidence estimator

s-jse committed:
Refactor: extract `ConfidenceEstimator` class

With two subclasses for two types of estimators:
raw score and decision tree based.

s-jse committed:
Fix LGTM alerts

s-jse committed:
Add confidence scores to server mode

s-jse committed:
Set save directory for calibration plots

s-jse committed:
Set default number of MC dropouts to 0

s-jse committed:
Fix bug: reverse almond task inherited from the wrong class

s-jse committed:
Fix bug

gcampax commented on genienlp/calibrate.py:
```diff
@@ -0,0 +1,510 @@
+from typing import Callable, Iterable, List, Tuple, Union
```
Missing copyright comment. Also, did you copy some of this code from somewhere? It looks "foreign" with the heavy functional style.

s-jse replied:
It's all mine! I'll add the usual copyright comment.

gcampax commented on genienlp/models/transformer_seq2seq.py:
```diff
@@ -159,3 +162,84 @@ def generate(self,
+        
+        # activate dropout layers
+        should_revert_to_eval = True
+        if mc_dropout:
```
This would be a bit more readable if "should_rever_to_eval" is set to True only in the if statement. Then later you don't need to check "mc_dropout"

s-jse replied:
`should_revert_to_eval` and `mc_dropout` are different things. It is possible that the caller has already put the model into `train()` mode, so after tha call is done, it should still be in `train()` mode.

gcampax replied:
If there is no mc_dropout, you don't call .train(), so you also don't need to call .eval() (that's why you check for both later).
I think it would be more readable as:
```
should_revert_to_eval = False
if mc_dropout and not self.training:
    self.train()
    should_rever_to_eval = True
```

so later you have:
```
if should_revert_to_eval:
    self.eval()
```

gcampax commented on genienlp/util.py:
```diff
@@ -102,6 +104,70 @@ def backward(self, s: str, occurrence: str):
+class ConfidenceOutput:
+    """
+    Contains all necessary features that are useful for calculating confidence of a single generated output
+    """
```
I find the naming of this class (and the flag "output_confidence") to be quite confusing. It doesn't contain the confidence, it contains the features to compute the confidence. So maybe "ConfidenceFeatures" ? I know it would be a bit confusing with the low-level features in confidence.py, but at least the confusion is contained in one file instead of being across many files and API boundaries.

gcampax commented on genienlp/validate.py:
```diff
@@ -81,21 +96,27 @@ def generate_with_model(model, data_iterator, numericalizer, task, args, predict
+        confidence_scores = confidence_estimator.estimate(confidences)
+        output = output + (confidence_scores, )
+
+    return output
```
I wonder if it's time to switch to a dataclass or namedtuple for this, it's a pretty big tuple already...

gcampax commented on genienlp/server.py:
```diff
@@ -168,6 +188,10 @@ def parse_argv(parser):
+    # for confidence estimation:
+    parser.add_argument('--calibrator_path', type=str, default=None, help='If provided, will be used to output confidence scores for each prediction.')
+    parser.add_argument("--mc_dropout", action='store_true', help='Monte Carlo dropout')
+    parser.add_argument("--mc_dropout_num", type=int, default=0, help='Number of samples to use for Monte Carlo dropout')
```
What is the correct configuration to use for MC dropout? Can we make it so the estimated confidence model stores how many MC dropout samples to take (or none if dropout features are not used)? Because Genie currently has no way to pass per-model flags to server.py

s-jse replied:
Fixed this.

gcampax commented on genienlp/server.py:
```diff
@@ -168,6 +188,10 @@ def parse_argv(parser):
     parser.add_argument('--locale', default='en', help='locale tag of the language to parse')
 
+    # for confidence estimation:
+    parser.add_argument('--calibrator_path', type=str, default=None, help='If provided, will be used to output confidence scores for each prediction.')
```
This should default to `--path` + /calibrator.pt or similar, and then you check if the file exists or not.

s-jse replied:
Fixed this.

gcampax reviewed (commented):
Do you have a PR for genie-k8s with the updated pipelines?
Also, is there a way with this to have examples that are known to be OOD? I guess it would be by marking them with some pseudo-ThingTalk that is never part of the training set, so the score is always zero? (That would be `$failed ;` in ThingTalk 2.0)

s-jse committed:
Add copyright statement

s-jse committed:
Rename confidence to confidence features

gcampax committed:
almond: fix evaluating confidence on OOD commands

OOD commands are annotated as "$failed ;" (the ThingTalk control
command for failure to parse / out-of-domain), which doesn't
start with "$dialogue ;". Note that we don't actually
train on those commands, only pass it to `genienlp predict`

s-jse committed:
Add output generation

Wraps around everything that
the generation function (`generate_with_model`) needs to output.

s-jse committed:
Add default calibrator path to server

s-jse committed:
Merge branch 'wip/confidence-scores-rebased' of https://github.com/stanford-oval/genienlp into wip/confidence-scores-rebased

gcampax committed:
Fix JSON serialization of confidence scores in server mode

s-jse committed:
Load calibration arguments when in server mode

s-jse committed:
Merge branch 'wip/confidence-scores-rebased' of https://github.com/stanford-oval/genienlp into wip/confidence-scores-rebased

s-jse committed:
Improve readability

s-jse committed:
Add tests for calibrator
